File: 2_Regression/2.2_Multiple_Linear_Regression/Self_Multiple_Linear_Regression.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Aug  8 14:01:12 2017

@author: ddxy
"""

#MUltiple Linear Regression

"""
Step 1: importing the libraries
"""

import numpy as np
import matplotlib.pyplot as plt

#panda's library is the best library to import and manage datasets 
import pandas as pd

"""
Step 2: importing the dataset
"""
#2.1: set a working directory by saving this file. The folder that holds this file
#will the be working directory

dataset = pd.read_csv('50_Startups.csv', header = 0)

#Replace some special characters in the header by '_' so that these headers
#can be read correctly in step 6 below
dataset.columns = [x.strip().replace(' ', '_') for x in dataset.columns]
dataset.columns = [x.strip().replace('&', '_') for x in dataset.columns]

#Print the headers
print(list(dataset))

#2.2: create a matrix of features/IVs
X = dataset.iloc[:,:-1].values #[:,:-1] means select all the rows and the 
                               #columns except the last column

#2.3: create a dependent variable vector
y = dataset.iloc[:,4].values

"""
Step 3: Convert categorical data to dummy variables
"""
#first convert the strings in the categorical column to integer coded column
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X = LabelEncoder()
X[:, 3] = labelencoder_X.fit_transform(X[:, 3])

#Second convert the integer coded column into dummy variables
onehotencoder = OneHotEncoder(categorical_features = [3])
X = onehotencoder.fit_transform(X).toarray()

#Avoiding the Dummy Variable Trap by elimilating the first column because we
#always want to omit 1 dummy variable to fulfill the "independence of errors"
#assumption

X = X[:, 1:]

"""
Step 4: Splitting the dataset into the training and test sets
"""
from sklearn.cross_validation import train_test_split
#random_state = 0 is to make everytime it gives the same result. it is like
#specifying a seed for random() in java
#practically, we do not need this so that every time the program will
#randomly select 20% of data for the test.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)


"""
Step 5: Create a Multi_linear_regression machine and let the machine learn
        In other words, create a model
"""
from sklearn.linear_model import LinearRegression
#create the MACHINE
regressor = LinearRegression()
#let the machine LEARN the data
regressor.fit(X_train, y_train)



"""
Step 6: Show statistical results of the model built on the training set
"""
from statsmodels.formula.api import ols

#split the dataset to training and test sets
train_set, test_set = train_test_split(dataset, test_size = 0.2, random_state = 0)

#create the formula string
tempHeaders = list(dataset)
formula = tempHeaders[-1] + " ~ "
for iv in range(0, len(tempHeaders) - 1):
    formula += tempHeaders[iv] + " + "

formula = formula[:-3]

#use ols to create the model
model = ols(formula, train_set).fit()
model.summary()


"""
Step 7: Predicting the Test set results
"""
y_pred = regressor.predict(X_test)



"""
Step 8: Building the optimal model using Backward Elimination
"""

File: 2_Regression/2.2_Multiple_Linear_Regression/Self_Multiple_Linear_Regression.R
#Multiple Linear Regression

#step 1: import the dataset
#set a working directory by going to the Files pane, find the correct folder, then go
#to More and click on "Set as Working Directory"
dataset = read.csv('50_Startups.csv')

#Step 2: Convert categorical data to dummy variables
dataset$State = factor(dataset$State,
                         levels = c('California', 'Florida', 'New York'), #c() means vector in R
                         labels = c(1, 2, 3))

#Step 3: Splitting the dataset into the Training set and Test set
#install caTools library by install.packages('caTools')
#Then, check caTools in the Packages pane. OR use
library(caTools)
set.seed(123) #set seed so that every time we get the same random result

#sample.split() will randomly select specified percentage of tuples to be the training
#set and label them as TRUE. The test set tuples are labeled as FALSE
split = sample.split(dataset$Profit, SplitRatio = 0.8)

#create traning set and test set based on the TRUE and FALSE label created in the 
#last step in dataset
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

#Feature scaling
#training_set[, 2:3] = scale(training_set[, 2:3])
#test_set[, 2:3] = scale(test_set[, 2:3])

#Step 4: create the model
#NOTE that R will automatically remove one of the dummy variables to fulfill
#the "independence of errors" assumption
regressor = lm(formula = Profit ~ ., # OR: formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
               data = training_set
               )
summary(regressor)
#show the coefficients of each terms in the model
coef(regressor)

#Step 5: Predicting the Test set results
y_pred = predict(regressor, newdata = test_set)






